### ISSUE TITLE
Typing

### ISSUE THREAD

**Issue by Luttik** [alex-sherman/unsync#15]
It would be great if the IDE could understand that an unsynced method returns an `Unfuture` rather than the usual data. And it would be even sweater if the `Unfuture` would be generic such that we know which value `.result()` returns.

**Comment by alex-sherman** [comment]
Agreed, that would actually be awesome. I'm looking into how that would be possible and I believe there is currently not support for decorators [modifying the return type](https://github.com/python/mypy/issues/3157) of their functions within the Python typing specification.

If there are any examples of this being done, I would be more than happy to look further into it, but for the time being I think there is nothing I can do to support this.

**Comment by Luttik** [comment]
The code in [this commit](https://github.com/Luttik/unsync/commit/ec2d07efd3293545729563082da7ec3efc7094b6) works if you use python 3.5 or high. Makes the issue less severe, since you can then type the response of your method as future manually. This way you only need to disallow type inspection at the point where you return the response of the method. And the IDE knows that `.result()` is a method of the response of `test_dict()` and that `result()` now returns a dict.

```
@unsync
def test_dict() -> Unfuture[dict]:
    # noinspection PyTypeChecker
    return {'x': 1}


def use_dict():
    _dict = test_dict().result()
    _dict.get('x')
```

**Comment by Luttik** [comment]
My message above however does have two issues.

1. You still have to use a noinspect somewhere.
2. It will not work for python < 3.5 (I guess there are options to fix this but I never build stuff for python < 3.7 so I don't have experience in this area)

**Comment by ekevoo** [comment]
The way I solve this problem is an assert isinstance after the call that returns unfuture. Not ideal either, but grants me full inspection.

**Comment by alex-sherman** [comment]
I think this library cannot target Python < 3.5 anyway because it uses the `async` keyword, so I'm OK with including things that require >=3.5. @Luttik  I actually really like the idea of supporting typing inside `Unfuture` at, if you open a pull request with that commit I'm happy to accept it unless you think there is some reason not to include that. It doesn't fully solve the problem but at least it's a step in the right direction.

**Comment by roee30** [comment]
Commenting on this:

> As far as we know it is not possible to change the return type of a method or function using a decorator.

Try this:
```python
import typing
if typing.TYPE_CHECKING:
    def unsync[**P, R](f: Callable[P, R]) -> Callable[P, Unfuture[R]]: ...
else:
    class unsync(metaclass=unsync_meta):
        # implementation
```
This is with Python 3.12 [type parameter syntax](https://peps.python.org/pep-0695/) but it can be also written in <3.12 syntax.

Another possibility that doesn't need `typing.TYPE_CHECKING` and might work is annotating `unsync.__call__` similarly.
